azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+            {"trigger": "_trig_cancel", "source": "disconnected", "dest": None},
+            {
+                "trigger": "_trig_cancel",
+                "source": "registering",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "waiting_to_poll",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "polling",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+        ]
+
+        def _on_transition_complete(event_data):
+            if not event_data.transition:
+                dest = "[no transition]"
+            else:
+                dest = event_data.transition.dest
+            logger.info(
+                "Transition complete.  Trigger=%s, Dest=%s, result=%s, error=%s",
+                event_data.event.name,
+                dest,
+                str(event_data.result),
+                str(event_data.error),
+            )
+
+        self._state_machine = Machine(
+            model=self,
+            states=states,
+            transitions=transitions,
+            initial="disconnected",
+            send_event=True,  # Use event_data structures to pass transition arguments
+            finalize_event=_on_transition_complete,
+            queued=True,
+        )
+
+    def register(self, callback=None):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("register called from polling machine")
+        self._register_callback = callback
+        self._trig_register()
+
+    def cancel(self, callback=None):
+        logger.info("cancel called from polling machine")
+        self._cancel_callback = callback
+        self._trig_cancel()
+
+    def _initialize_register(self, event_data):
+        logger.info("Preparing for registration process.")
+        # Choice: Sending action here as the polling/dps state machine must figure out
+        # in what order action must be sent. The request response is only responsible for
+        # requesting and responding. However the actions can be taken down to the
+        # the request response machine as well.
+        subscribe_action = SubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, qos=1, callback=self.on_subscribe_completed
+        )
+        # The request response takes a tuple of actions
+        # It will call state based provider connect which
+        # It will add subscribe to pending action queue
+        # it will execute connect and then execute subscribe
+        # TODO : Should I have a waiting callback or a connack on which something happens ?
+        self._request_response_provider.connect(
+            self.on_connect_completed, post_connect_actions=[subscribe_action]
```

Comment:
>post_connect_actions=[subscribe_action] [](start = 39, length = 39)

I'm not sure about this design. It seems overly complex to be injecting your own actions to trigger after the operation you call.

1) This presumably creates a LOT more edge cases
2) It relies on a knowledge of the implementation of the transport/request_response_provider, which I don't think should be taken for granted across layers of code
3) It offers an additional way to do something that you already can do - Subscribe and Connect are offered as two separate APIs, but this chaining option makes that irrelevant.

I'd like to revisit this when we start bringing the codebase together, I need to be convinced that this is both necessary, and the right solution to the problem.

Reply:
changed to just subscribe. not needed. 
yes it does require some knowledge of transport.
however i felt at the time of doing this that this is happening at the polling machine level and not at client so , depending on the type of feature we are in that level may decide to do certain actions after connection and prior disconnection..that point is moot at this time coz of pipeline where the stages will determine what happens.

---
In reply to: [277816544](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277816544) [](ancestors = 277816544)